libfreerdp/core/transport.c

```diff
@@ -59,17 +59,130 @@
+	krb5_principal principal = NULL;
+	char address[256];
+	krb5_ccache ccache;
+	krb5_init_creds_context ctx = NULL;
+	_snprintf(address, sizeof(address), "%s@%s", username, domain);
+
+	/* Create a krb5 library context */
+	if ((ret = krb5_init_context(&context)) != 0)
+		WLog_ERR(TAG, "krb5_init_context failed with error %d", (int)ret);
+	else if ((ret = krb5_parse_name_flags(context, address, 0, &principal)) != 0)
+		WLog_ERR(TAG, "krb5_parse_name_flags failed with error %d", (int)ret);
+	/* Find a credential cache with a specified client principal */
+	else if ((ret = krb5_cc_cache_match(context, principal, &ccache)) != 0)
+	{
+		if ((ret = krb5_cc_default(context, &ccache)) != 0)
+			WLog_ERR(TAG, "krb5 failed to resolve credentials cache with error %d", (int)ret);
+	}
+
+	if (ret != KRB5KDC_ERR_NONE)
+		goto out;
+	/* Create a context for acquiring initial credentials */
+	else if ((ret = krb5_init_creds_init(context, principal, NULL, NULL, 0, NULL, &ctx)) != 0)
+	{
+		WLog_WARN(TAG, "krb5_init_creds_init returned error %d", (int)ret);
+		goto out;
+	}
+	/* Set a password for acquiring initial credentials */
+	else if ((ret = krb5_init_creds_set_password(context, ctx, passwd)) != 0)
+	{
+		WLog_WARN(TAG, "krb5_init_creds_set_password returned error %d", ret);
+		goto out;
+	}
+
+	/* Acquire credentials using an initial credential context */
+	ret = krb5_init_creds_get(context, ctx);
+out:
+
+	switch (ret)
+	{
+		case KRB5KDC_ERR_NONE:
+			break;
+
+		case KRB5_KDC_UNREACH:
+		case KRB5KRB_AP_ERR_BAD_INTEGRITY:
+		case KRB5KRB_AP_ERR_MODIFIED:
+		case KRB5KDC_ERR_PREAUTH_FAILED:
+		case KRB5_GET_IN_TKT_LOOP:
+			WLog_WARN(TAG, "krb5_init_creds_get: Password incorrect");
+			ret = FREERDP_ERROR_AUTHENTICATION_FAILED;
+			break;
+
+		case KRB5KDC_ERR_KEY_EXP:
+			WLog_WARN(TAG, "krb5_init_creds_get: Password has expired");
+			ret = FREERDP_ERROR_CONNECT_PASSWORD_EXPIRED;
+			break;
+
+		case KRB5KDC_ERR_CLIENT_REVOKED:
+			ret = FREERDP_ERROR_CONNECT_PASSWORD_CERTAINLY_EXPIRED;
+			break;
+
+		case KRB5KDC_ERR_POLICY:
+			ret = FREERDP_ERROR_INSUFFICIENT_PRIVILEGES;
+			break;
+
+		default:
+			WLog_WARN(TAG, "krb5_init_creds_get");
+			ret = FREERDP_ERROR_CONNECT_TRANSPORT_FAILED;
+			break;
+	}
+
+	if (ctx)
+		krb5_init_creds_free(context, ctx);
+
+	krb5_free_context(context);
+	return ret;
+}
+#endif /* WITH_KRB5 */
 
 static void transport_ssl_cb(SSL* ssl, int where, int ret)
 {
 	rdpTransport* transport;
 
-	if ((where | SSL_CB_ALERT) && (ret == 561))
+	if (where | SSL_CB_ALERT)
 	{
-		transport = (rdpTransport*) SSL_get_app_data(ssl);
+		if (ret == 561)
+		{
+			transport = (rdpTransport*) SSL_get_app_data(ssl);
 
-		if (!freerdp_get_last_error(transport->context))
-			freerdp_set_last_error(transport->context, FREERDP_ERROR_AUTHENTICATION_FAILED);
+			if (!freerdp_get_last_error(transport->context))
+			{
+				freerdp_set_last_error(transport->context, FREERDP_ERROR_AUTHENTICATION_FAILED);
+			}
+		}
+		else if (ret == 592)
```

Comment:
Looking at https://linux.die.net/man/3/ssl_set_info_callback the `ret` value can be interpreted with `ssl_alert_type_string`. This function return F (as Fatal)  

`case SSL3_AL_FATAL:
        return "F";`